server/src/message_cache.rs

```diff
@@ -0,0 +1,631 @@
+        let mut removed_messages: Vec<Envelope> = Vec::new();
+
+        for guid in message_guids {
+            let message_id: Option<String> = cmd("HGET")
+                .arg(queue_metadata_key.clone())
+                .arg(guid.clone())
+                .query_async(&mut conn)
+                .await?;
+
+            if let Some(msg_id) = message_id.clone() {
+                // retrieving the message
+                let envelope: Option<Vec<Vec<u8>>> = cmd("ZRANGE")
+                    .arg(queue_key.clone())
+                    .arg(msg_id.clone())
+                    .arg(msg_id.clone())
+                    .arg("BYSCORE")
+                    .arg("LIMIT")
+                    .arg(0)
+                    .arg(1)
+                    .query_async(&mut conn)
+                    .await?;
+
+                // delete the message
+                cmd("ZREMRANGEBYSCORE")
+                    .arg(queue_key.clone())
+                    .arg(msg_id.clone())
+                    .arg(msg_id.clone())
+                    .query_async::<()>(&mut conn)
+                    .await?;
+
+                // delete the guid from the cache
+                cmd("HDEL")
+                    .arg(queue_metadata_key.clone())
+                    .arg(guid.clone())
+                    .query_async::<()>(&mut conn)
+                    .await?;
+
+                if let Some(envel) = envelope {
+                    removed_messages.push(bincode::deserialize(&envel[0])?);
+                }
+            }
+        }
+
+        if cmd("ZCARD")
+            .arg(queue_key.clone())
+            .query_async::<u64>(&mut conn)
+            .await?
+            == 0
+        {
+            cmd("DEL")
+                .arg(queue_key.clone())
+                .query_async::<()>(&mut conn)
+                .await?;
+
+            cmd("DEL")
+                .arg(queue_metadata_key.clone())
+                .query_async::<()>(&mut conn)
+                .await?;
+
+            cmd("ZREM")
+                .arg(queue_total_index_key.clone())
+                .arg(queue_key.clone())
+                .query_async::<()>(&mut conn)
+                .await?;
+        }
+
+        Ok(removed_messages)
+    }
+
+    pub async fn has_messages(&self, user_id: String, device_id: u32) -> Result<bool> {
+        let mut conn = self.pool.get().await.unwrap();
+
+        let msg_count = cmd("ZCARD")
+            .arg(MessageCache::get_message_queue_key(user_id, device_id))
+            .query_async::<u32>(&mut conn)
+            .await?;
+
+        Ok(msg_count > 0)
+    }
+
+    pub async fn get_all_messages(&self, user_id: String, device_id: u32) -> Vec<Envelope> {
+        let messages = self.get_items(user_id.clone(), device_id.clone(), -1).await;
+
+        if (messages.is_empty()) {
+            return Vec::new();
+        }
+        let mut envelopes = Vec::new();
+
+        for i in (0..messages.len()).step_by(2) {
```

Comment:
Can you make a comment on how this actually works. As you said messages are pairs as env, id and you only want the env